### PR TITLE
DATACOUCH-595 - Add cas to replace options if present on entity.

### PR DESCRIPTION
**Pull request by mikereiche**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplateSupport.java [hunk 1]
```diff
@@ -96,6 +96,18 @@ public void applyUpdatedCas(final Object entity, final long cas) {
 		}
 	}
 
+	public long getCas(final Object entity) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplateSupport.java:99]: Nit: We typically don't use `final` keywords within methods.

File: src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplateSupport.java [hunk 2]
```diff
@@ -96,6 +96,18 @@ public void applyUpdatedCas(final Object entity, final long cas) {
...
+		final ConvertingPropertyAccessor<Object> accessor = getPropertyAccessor(entity);
+		final CouchbasePersistentEntity<?> persistentEntity = mappingContext.getRequiredPersistentEntity(entity.getClass());
+		final CouchbasePersistentProperty versionProperty = persistentEntity.getVersionProperty();
+
+		long cas = 0;
+		if (versionProperty != null) {
+			cas = (long) accessor.getProperty(versionProperty);
```
Comment (mp911de) [inline src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplateSupport.java:106]: Probably safer to have an instanceof check for `Number` as domain models may contain `BigInteger` and other numeric types.

File: src/test/java/org/springframework/data/couchbase/core/CouchbaseTemplateKeyValueIntegrationTests.java [hunk 3]
```diff
@@ -74,6 +74,12 @@ void upsertAndFindById() {
 		User modified = couchbaseTemplate.upsertById(User.class).one(user);
 		assertEquals(user, modified);
 
+		modified = couchbaseTemplate.replaceById(User.class).one(user);
+		assertEquals(user, modified);
```
Comment (mp911de) [inline src/test/java/org/springframework/data/couchbase/core/CouchbaseTemplateKeyValueIntegrationTests.java:78]: On a related note, Spring Data has switched to AssertJ which makes assertions a bit more convenient.

File: src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java [hunk 4]
```diff
@@ -68,7 +68,7 @@ public ReactiveReplaceByIdOperationSupport(final ReactiveCouchbaseTemplate templ
 			return Mono.just(object).flatMap(o -> {
 				CouchbaseDocument converted = template.support().encodeEntity(o);
 				return template.getCollection(collection).reactive()
-						.replace(converted.getId(), converted.export(), buildReplaceOptions()).map(result -> {
+						.replace(converted.getId(), converted.export(), buildReplaceOptions(object)).map(result -> {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java:71]: nit: this should be o, not object, right?

Comment (daschl) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java:71]: actually I wonder if the code from below is put up here and only the long cas is put into the buildReplaceOptions

File: src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java [hunk 5]
```diff
@@ -86,13 +86,17 @@ public ReactiveReplaceByIdOperationSupport(final ReactiveCouchbaseTemplate templ
...
-		private ReplaceOptions buildReplaceOptions() {
+		private ReplaceOptions buildReplaceOptions(T object) {
 			final ReplaceOptions options = ReplaceOptions.replaceOptions();
 			if (persistTo != PersistTo.NONE || replicateTo != ReplicateTo.NONE) {
 				options.durability(persistTo, replicateTo);
 			} else if (durabilityLevel != DurabilityLevel.NONE) {
 				options.durability(durabilityLevel);
 			}
+			long cas = template.support().getCas(object);
+			if (cas != 0) {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java:97]: 0 cas == not present so if 0 is passed in its fine

File: src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java (hunk 4 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java:71]: Yes. should be o.  Fixed.

File: src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java (hunk 5 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java:97]: fixed

File: src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplateSupport.java (hunk 2 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplateSupport.java:106]: fixed

File: src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java (hunk 4 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveReplaceByIdOperationSupport.java]: The durability also needs to go in the options.

**Review by daschl — APPROVED** [review]

**Comment by daschl** [comment]
@mikereiche also I think this should be backported to 4.0 ?

**Pull request closed without merging**